README
"NRG Ljubljana" is a flexible framework for performing large scale numerical
renormalization group (NRG) calculations for quantum impurity problems. It
is highly extensible without sacrificing numerical efficiency.

Copyright (C) 2006-2014 Rok Zitko


Framework "NRG Ljubljana" is a set of interrelated computer codes for
performing numerical renormalization group (NRG) calculations for quantum
impurity problems, described by models such as the Kondo exchange (s-d)
model or the Anderson single impurity model, and their multi-impurity and
multi-channel generalizations. It also contains a number of tools for
analyzing results (thermodynamic properties, such as magnetic and charge
susceptibility, entropy and heat capacity; expectation values of arbitrary
operators; spectral functions). It is user friendly, in the sense that it is
easy to set up new types of problems (Hamiltonians, perturbation terms,
etc.) and the output is formatted and annotated for easy interpretation,
parsing and plotting. It efficiently handles problems with different
symmetries, such as spin SU(2) symmetry, charge SU(2) symmetry, Z_2
reflection symmetry (parity), etc.

To achieve a high degree of flexibility without sacrificing numerical
efficiency, "NRG Ljubljana" is composed of a hierarchy of modules:
high level modules are written in a mixture of functional and
procedural Mathematica code, while the low level numerically intensive
parts are programmed in the object oriented approach in the C++
language. The foundation of the framework is a Mathematica package for
performing calculations with non-commutative second quantisation
operators, SNEG. Next layer is a Mathematica program which defines the
Hamiltonian, the basis of states, and the physical operators of
interest: with the help of SNEG, Hamiltonian and operators can be
defined using the familiar second-quantization expressions. This
program performs the diagonalization of the initial Hamiltonian and
prepares the input for the NRG iteration proper.

For efficiency, NRG iteration is performed by a separate C++ program:
for a typical problem, most of the time (90%) is spent in the LAPACK
dsyev routine which solves the eigenvalue problem. There is very
little housekeeping overhead due to the tasks required by the NRG
iteration; "NRG Ljubljana" is thus suitable for performing large scale
NRG calculations on computer clusters.



1. Features

- all parameters and model definitions (!) configurable at run-time, no need
  to recompile for different runs
- support for a large number of different symmetry types
- flexible truncation schemes (energy cut-off truncation, avoidance
  of trunction within gaps, etc.)
- density-matrix NRG (DM-NRG) for all symmetry types
- full-density matrix (FD-NRG) for all symmetry types
- FDM calculation of expectation values at finite temperatures
- sum-rule conserving spectral function calculations (CFS approach)
- various spectral-function broadening schemes & stand-alone tools
- self-energy trick calculations 
- arbitrary number of channels (in practice 1,2,3)
- calculations with real and complex numbers
- support for superconducting bands, spin-polarized bands, etc.
- support for global operators (i.e., operators defined on the
  Wilson chain sites)
- calculation of temperature-dependent conductance, G(T)
- flexibility in definition of the Hamiltonian, which is simply
  written using a second-quantization operator expression
- automatic exact diagonalisation of the initial Hamiltonian with
  automagic generation of the (Q,S) (or some other) basis
- flexibility in the choice of operators whose thermodynamical averages
  are computed, they can be expressed using operators of second
  quantization
- dynamic spin susceptibility, dynamic charge susceptibility, etc.
  It is possible to compute arbitrary spectral functions for any
  pair of local operators.
- multiple logarithmic discretization schemes (Wilson/Krishnamurthy,
  Yoshida/Whitaker/Oliveira, Campo/Oliveira, ODE scheme)
- support for non-flat bands (i.e. cosine band that arrises from
  tight-binding description of leads in quantum transport problems,
  or arbitrary hybridization as needed in DMFT), see dmft.m
- heavily commented code with reference to original publications of
  formulas and approaches used
- object-oriented code for easy maintenance and expandability
- FAST!: very high numeric efficiency with optimized-for loops in the most
  numerically demanding parts of the code (chiefly the recalculation
  of irreducible matrix elements of operators)
- N/N+2 trick (patching) in calculation of spectral densities
- amount of verbosity in log files is configurable
- automatic recording of time elapsed in various parts of the program
- monitoring of memory use
- formated output files for easy interpretation of the results
  (for example, headers with parameters and heading lines in tables
  with operator names)
- internal consistency checks, assertions, parameter compatibility
  and reasonableness checks that reduce the possibility of undetected
  bugs
  

2. Documentation

Presently, there is no up-to-date documentation. Sorry.


3. License

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

The full text of the GPL General Public License can be found
in file LICENSE.


4. Compiling and dependencies

NRG Ljubljana is very portable and it should work without any modification
on any modern Linux distribution and, with some tweaking, on any Unix or
Unix-like operating system with a good standards-compliant C++ compiler. It
has been reported to me that it can also be compiled under Windows.

The following libraries are required to compile the C++ part of the
NRG code:

    * LAPACK and BLAS linear algebra libraries
    * boost C++ libraries

Due to the heavy use of template metaprogramming in boost libraries, a
high-quality standards-compliant C++ compiler must be used. 

Wolfram Reasearch Mathematica must be installed for running the
Mathematica part of the NRG code. Versions 5.0, 5.1, 5.2, 6.0, 7.0 and
8.0 have been tested. Mathematica is only required for the
initialization of the problem (basis construction, diagonalisation of
the initial Hamiltonian, transformations of the operator matrices,
etc.) which is relatively fast. When "NRG Ljubljana" is used on a
cluster, it is therefore sufficient to have Mathematica installed on a
single computer (for example on the cluster host computer), while the
numerically demanding (C++) part of the program can be ran on the
cluster nodes.

Usually, the compilation consists of configuring and compiling NRG:

./configure --prefix=$HOME --with-tools
make

(The tools are optional: you might not need them. In this case
do not use the "--with-tools" option.)

5. "Installation"

To install "NRG Ljubljana", copy the executables to an appropriate location.
Furthermore, the Mathematica scripts *.m must be accessible; you may simply
copy them to the directory from which the NRG code will be started.

You can also do a "make install". Note, however, that the scripts from
scripts/ and tools from tools/ will end up in the installation bin/
directory.


6. Contributing to "NRG Ljubljana"

If you make improvements to "NRG Ljubljana", you are encouraged to
share them with other users. Bug reports (and fixes) are very welcome
as well.  The contact information is in the next section.


7. Contact information:

"NRG Ljubljana" home-page: http://nrgljubljana.ijs.si/

Rok Zitko
"Jozef Stefan" Institute
F1 - Theoretical physics
Jamova 39
SI-1000 Ljubljana
Slovenia

rok.zitko@ijs.si


8. Acknowledgements 

The development of the "NRG Ljubljana" framework started during author's PhD
studies at the Faculty for mathematics and physics of the University of
Ljubljana, and the "Jozef Stefan" Institute, Ljubljana, Slovenia.
Discussions and collaboration with prof. Janez Bonca, prof. Anton Ramsak,
dr. Jernej Mravlje and dr. Tomaz Rejec from the F1, Theoretical Physics
department are acknowledged. I'm also grateful to prof. Thomas Pruschke,
Robert Peters and Oliver Bodensiek from the University in Goettingen for many
very fruitful discussions. I thank Marcus Greger from the University
in Augsburg for contributing optimized routines for the spectral
function calculation.




NEWS
1.0 - first public release (Jan 2007)
1.3 - public release (Jul 2007)
2.0.6 - public release (Nov 2009)
2.2.0 - public release (March 2011)
2.4 - major code cleanup; breaks backwards compatibility (2013, 2014)


src/matrix.cc
// "NRG Ljubljana" - Numerical renormalization group for multiple
// impurities and an arbitrary number of channels
//
// matrix.cc - Symmetry dependent code for Hamiltonian matrix generation
//
// Copyright (C) 2009-2016 Rok Zitko
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation; either version 2 of the License, or
//   (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License
//   along with this program; if not, write to the Free Software
//   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
//
//   Contact information:
//   Rok Zitko
//   F5 - Condensed matter physics
//   "Jozef Stefan" Institute
//   Jamova 39
//   SI-1000 Ljubljana
//   Slovenia
//
//   rok.zitko@ijs.si

#ifndef _matrix_cc_
#define _matrix_cc_

// TO DO: remove P::, move all the required information to the
// symmetry class where it belongs.

// Check i,j,ch parameters of *diag_function().
void check_ijch(size_t i, size_t j, size_t ch)
{
   allowed_block_index(i);
   allowed_block_index(j);
   allowed_channel(ch);
}

/* +++ Construct an offdiagonal part of the Hamiltonian. +++

 i,j - indexes of matrix blocks that constitue the total Hamiltonian matrix
       in the given invariant subspace. Not all possible blocks do contribute.
 ch - contribution of which channel is being considered? 0,1,..
 factor0 - the coefficient which multiplies the irreducible matrix
           elements. This coefficient takes into account the multiplicities.
           For simple symmetries, this factor is +-1, where the sign is
           related to the fermionic anticommutations that are required to
           shift around the creation/annihilation operators that appear in the
           hopping operator.

 NOTE: the offdiagonal part depends on xi(N), while zeta(N) affect the
 diagonal part of the Hamiltonian matrix! 
*/

bool offdiag_contributes(size_t i, size_t j, size_t ch,
                         const Rmaxvals &qq)
{
  check_ijch(i, j, ch);
  my_assert(i != j);
  const size_t size1 = qq.rmax(i);
  const size_t size2 = qq.rmax(j);
  // Here we test if the block (i,j) exists at all. If it doesn't, we
  // immediately return, otherwise the following assertion will trigger a
  // false error message and abort the execution.
  const bool contributes = (size1 > 0) && (size2 > 0);
  return contributes;
}

// i,j - indexes of the out-of-diagonal matrix block that we are
//       constructing
// ch,fnr - channel and extra index to locate the correct block of
//          the <||f||> matrix (irreducible matrix elements)
// factor - appropriate factor (includes the Clebsch-Gordan coefficients
//          as well as the xi hopping parameter)
// h - matrix being built
// qq - matrix dimensions data (to determine the position in the matrix)
// In - In[i] and In[j] are the invariant subspaces of required <||f||>
//      matrix elements
void offdiag_build(size_t i, size_t j,
                   size_t ch, // channel number
                   size_t fnr, // extra index for <||f||>, usually 0
                   t_matel factor, // may be complex (in principle)
                   Matrix &h,
                   const Rmaxvals &qq,
                   const InvarVec &In)
{
  // We are building the upper triangular part of the symmetric Hamiltonian
  // matrix! Thus usually i > j. If not, we must conjugate transpose
  // the contribution!
  const bool transpose = i > j;
  const size_t begin1 = qq.offset(i);
  const size_t begin2 = qq.offset(j);
  const size_t size1 = qq.rmax(i);
  const size_t size2 = qq.rmax(j);
  // already checked in offdiag_contributes(), but we do it
  // again out of paranoia...
  my_assert(size1 && size2);
  // < In[i] r | f^\dag | In[j] r' >
  const Twoinvar II = make_pair(In[i], In[j]);
  if (!isfinite(factor)) {
     cout << "offdiag_function() critical error: factor is not finite." << endl;
     cout << "i=" << i << " j=" << j << " ch=" << ch << " fnr=" << fnr << " factor=" << factor << endl;
     cout << "source subspaces II=" << II << endl;
     exit(1);
  }
  const t_matel factor_scaled = factor/SCALE(STAT::N+1);
#define F_EPSILON 1e-8
  if (abs(factor_scaled) < F_EPSILON)
     return; // Doesn't contribute after all. Factors are usually order 1.
  const size_t cnt = a.opch[ch][fnr].count(II);
  if (cnt != 1) {
     cout << "offdiag_function() critical error: <||f||> subspace does not exist." << endl;
     cout << "i=" << i << " j=" << j << " ch=" << ch << " fnr=" << fnr << " factor_scaled=" << factor_scaled << endl;
     cout << "II=" << II << " cnt=" << cnt << endl;
     exit(1);
  }
  my_assert (size1 == a.opch[ch][fnr][II].size1());
  my_assert (size2 == a.opch[ch][fnr][II].size2());
  nrglog('i', "offdiag i=" << i << " j=" << j << 
	 " factor_scaled=" << factor_scaled << " II=" << II);
  if (transpose) {
     matrix_range<Matrix>
       hsub(h, range(begin2, begin2+size2), range(begin1, begin1+size1));
     noalias(hsub) += CONJ_ME(factor_scaled) * herm(a.opch[ch][fnr][II]);
  } else {
     matrix_range<Matrix>
       hsub(h, range(begin1, begin1+size1), range(begin2, begin2+size2));
     noalias(hsub) += factor_scaled * a.opch[ch][fnr][II];
  }
}

void offdiag_function(size_t i, size_t j,
                      size_t ch, // channel number
                      size_t fnr, // extra index for <||f||>, usually 0
		      t_matel factor, // may be complex (in principle)
		      Matrix &h,
		      const Rmaxvals &qq,
		      const InvarVec &In)
{
   const bool contributes = offdiag_contributes(i, j, ch, qq);
   if (contributes)
      offdiag_build(i, j, ch, fnr, factor, h, qq, In);
}

// TRICK: this macro is used when the evaluation of 'factor' could
// lead to a seg fault. Here we check if the submatrix exists before
// the expression in 'factor' is evaluated.

#define offdiag_macro(i, j, ch, fnr, factor, h, qq, In) \
   { const bool contributes = offdiag_contributes(i, j, ch, qq); \
     if (contributes) { offdiag_build(i, j, ch, fnr, factor, h, qq, In); }; }

/* +++ Shift the diagonal matrix elements by the number of electrons
 multiplied by the required constant(s) zeta. +++

 'number' is the number of electrons for channel 'ch' in invariant
 subspaces indexed by 'i'. Note that 'number' is a floating point
 number: this is required, for example, in the LR symmetric basis
 sets.

 NOTE: for problems where a given invariant subspace does not
 correspond to a fixed number of added electrons, a generalized
 routine should be used. 
*/
void diag_function(size_t i, size_t ch, double number,
                   t_coef sc_zeta,
                   Matrix &h,
                   const Rmaxvals &qq)
{
  allowed_block_index(i);
  my_assert(number >= 0.0 && number <= 14.0);
  const size_t begin1 = qq.offset(i);
  const size_t size1 = qq.rmax(i);
  // For convenience we subtract the average site occupancy.
  const double avgoccup = ((double)P::spin)/2; // multiplicity divided by 2
  /* Energy shift of the diagonal matrix elements in the NRG Hamiltonian.
   WARNING: for N=0, we are not adding the first site of the Wilson chain
   (indexed as 0), but the second one (indexed as 1). Therefore the
   appropriate zeta is not zeta(0), but zeta(1). zeta(0) is the shift
   applied to the f[0] orbital in initial.m !!! */
  const t_coef shift = sc_zeta * (number-avgoccup) / SCALE(STAT::N+1);
  nrglog('i', "diag i=" << i << " shift=" << shift);
  for (size_t j = begin1; j < begin1+size1; j++)
    h(j, j) += shift;
}


// Compare with diag_function()
void diag_function_half(size_t i, size_t ch, double number,
                        t_matel sc_zeta,
                        Matrix &h,
                        const Rmaxvals &qq)
{
  allowed_block_index(i);
  my_assert(0.0 <= number && number <= P::spin);
  const size_t begin1 = qq.offset(i);
  const size_t size1 = qq.rmax(i);
  // For convenience we subtract the average site occupancy.
  const double avgoccup = ((double)P::spin)/2; // multiplicity divided by 2
  // avgoccup is divided by a further factor of 2 compared
  // to diag_function() above!
  const t_matel shift = sc_zeta * (number-avgoccup/2) / SCALE(STAT::N+1);
  nrglog('i', "diag_half i=" << i << " shift=" << shift);
  for (size_t j = begin1; j < begin1+size1; j++)
    h(j, j) += shift;
}

// Compare with diag_function() above.
void spinz_function(size_t i, size_t j,
                    size_t ch, t_matel spinz,
                    Matrix &h, const Rmaxvals &qq)
{
  check_ijch(i, j, ch);
  my_assert(i == j);

  // compare with the ISOSPINX macro
   const t_matel shift = spinz * double(P::globalB) / SCALE(STAT::N+1);

  const size_t begin1 = qq.offset(i);
  const size_t size1 = qq.rmax(i);
  nrglog('i', "spinz i=" << i << " shift=" << shift);
  for (size_t j = begin1; j < begin1+size1; j++)
    h(j, j) += shift;
}

void spinx_function(size_t i, size_t j,
                    size_t ch, t_matel spinx,
                    Matrix &h, const Rmaxvals &qq)
{
  check_ijch(i, j, ch);
  const t_matel shift = spinx * double(P::globalBx) / SCALE(STAT::N+1);
  if (i>j)
     return; // only upper triangular part
  size_t begin1 = qq.offset(i);
  size_t size1 = qq.rmax(i);
  size_t begin2 = qq.offset(j);
  size_t size2 = qq.rmax(j);
  bool contributes = size1 * size2;
  if (!contributes)
    return;
  my_assert(size1 == size2);
  nrglog('i', "spinx i=" << i << " shift=" << shift);
  for (size_t l = 0; l < size1; l++)
    h(begin1 + l, begin2 + l) += shift;
}

// +++ Shift the offdiagonal matrix elements by factor. +++

void diag_offdiag_function(size_t i, size_t j, size_t chin, t_matel factor,
                           Matrix &h,
                           const Rmaxvals &qq)
{
   check_ijch(i, j, chin);
   if (i>j)
     return; // only upper triangular part
   size_t begin1 = qq.offset(i);
   size_t size1 = qq.rmax(i);
   size_t begin2 = qq.offset(j);
   size_t size2 = qq.rmax(j);
   bool contributes = size1 * size2;
   if (!contributes)
     return;
   my_assert(size1 == size2);
   const t_matel factor_scaled = factor/SCALE(STAT::N+1);
  nrglog('i', "diag_offdiag i=" << i << " j=" << j << 
	 " factor_scaled=" << factor_scaled);
   for (size_t l = 0; l < size1; l++)
     h(begin1 + l, begin2 + l) += factor_scaled;
}

#endif // _matrix_cc_
